### PR TITLE
Make shutdown of `ReplicatedMergeTree` tables more soft

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce a table setting `wait_for_unique_parts_send_before_shutdown_ms` which specify the amount of time replica will wait before closing interserver handler for replicated sends. Also fix inconsistency with shutdown of tables and interserver handlers: now server shutdown tables first and only after it shut down interserver handlers.

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
This is an automated comment for commit baee73fd96d1b1974ac7ec637c3b22c4f63a27a4 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51851/baee73fd96d1b1974ac7ec637c3b22c4f63a27a4/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -4694,14 +4831,58 @@ void StorageReplicatedMergeTree::startupImpl(bool from_attach_thread)
...
+
+    /// Cancel fetches, merges and mutations to force the queue_task to finish ASAP.
+    fetcher.blocker.cancelForever();
+    merger_mutator.merges_blocker.cancelForever();
+    parts_mover.moves_blocker.cancelForever();
+    mutations_finalizing_task->deactivate();
+    stopBeingLeader();
+
+    if (attach_thread)
+        attach_thread->shutdown();
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:4850]: Hm, looks like we only need this (not sure about outdated parts and attach_thread), everything else can be stopped in normal `shutdown`.

However, `mutations_finalizing_task` is suspicious: we deactivate it before stopping restarting_thread, but restarting_thread can restart it: 
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L144

(we already had this issue before)

File: programs/server/Server.cpp [hunk 2]
```diff
@@ -1511,10 +1520,13 @@ try
         {
             LOG_DEBUG(log, "Waiting for current connections to servers for tables to finish.");
             size_t current_connections = 0;
-            for (auto & server : servers_to_start_before_tables)
             {
-                server.stop();
-                current_connections += server.currentConnections();
+                std::lock_guard lock(servers_lock);
```
[the remark this answers is not in the document]
Comment (alesapin) [reply programs/server/Server.cpp:1524]: Yep, agree.

**Comment by alesapin** [comment]
Possible deadlock traces: https://pastila.nl/?0006a09d/8960894b9787bcf5ffd90ee2956e4eae. At first glance it doesn't look like something related to changes, I'll try to investigate.

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -3974,6 +3979,146 @@ String StorageReplicatedMergeTree::findReplicaHavingPart(const String & part_nam
...
+
+        if (!found)
+        {
+            LOG_TRACE(log_, "Part not {} found on other replicas", part);
+            our_unique_parts.emplace(MergeTreePartInfo::fromPartName(part, format_version_));
+        }
+    }
+
+    if (!our_parts.empty() && our_unique_parts.empty())
+        LOG_TRACE(log_, "All parts found on replica");
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:4117]:
```suggestion
        LOG_TRACE(log_, "All parts found on replicas");
```

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 4]
```diff
@@ -4740,6 +4893,28 @@ void StorageReplicatedMergeTree::startupImpl(bool from_attach_thread)
...
+    fetcher.blocker.cancelForever();
+    merger_mutator.merges_blocker.cancelForever();
+    parts_mover.moves_blocker.cancelForever();
+    mutations_finalizing_task->deactivate();
+    stopBeingLeader();
+
+    if (attach_thread)
+        attach_thread->shutdown();
+
+    restarting_thread.shutdown(/* part_of_full_shutdown */true);
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:4912]: There's still a race condition between `mutations_finalizing_task->deactivate()` and `mutations_finalizing_task->activateAndSchedule()` in `ReplicatedMergeTreeRestartingThread::runImpl`

File: src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp [hunk 5]
```diff
@@ -339,10 +339,15 @@ void ReplicatedMergeTreeRestartingThread::shutdown(bool part_of_full_shutdown)
...
+    /// Explicitly set the event, because the restarting thread will not set it again
+    if (part_of_full_shutdown)
+        storage.startup_event.set();
+
     LOG_TRACE(log, "Restarting thread finished");
 
-    /// Stop other tasks.
-    partialShutdown(part_of_full_shutdown);
+    setReadonly(part_of_full_shutdown);
+
```
Comment (tavplubix) [inline src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp:350]: Looks like `setReadonly` will be called twice: from `ReplicatedMergeTreeRestartingThread::shutdown` from `flushAndPrepareForShutdown` and then again from `ReplicatedMergeTreeRestartingThread::partialShutdown` from `StorageReplicatedMergeTree::shutdown`. On the second call, `became_readonly` will be false here:
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L353-L355

So we will decrement the metric:
https://github.com/ClickHouse/ClickHouse/blob/2924837c7e6030058b87b9a65a06276543430cfd/src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp#L360-L365

**Review by tavplubix — APPROVED** [review]

**Review by alesapin — COMMENTED** [review]
Great code, will solve all possible issues🚀

**Comment by alesapin** [comment]
Real data race https://pastila.nl/?00105a7f/2e68bad235875ae48a49c42ea514bae2

**Comment by alesapin** [comment]
Trace from hung check https://pastila.nl/?0006a09d/bb7ff055e08a765a6e7801a786075044

**Review by alesapin — COMMENTED** [review]
That is it, it's a final version

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 6]
```diff
@@ -4740,6 +4900,36 @@ void StorageReplicatedMergeTree::startupImpl(bool from_attach_thread)
...
+            attach_thread->shutdown();
+
+        restarting_thread.shutdown(/* part_of_full_shutdown */true);
+        /// Explicetly set the event, because the restarting thread will not set it again
+        startup_event.set();
+        shutdown_deadline.emplace(std::chrono::system_clock::now() + std::chrono::milliseconds(settings_ptr->wait_for_unique_parts_send_before_shutdown_ms.totalMilliseconds()));
+    }
+    catch (...)
+    {
+        /// Don't wait anything in case of inproper prepare for shutdown
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:4927]:
```suggestion
        /// Don't wait anything in case of improper prepare for shutdown
```

**Comment by nikitamikhaylov** [comment]
I think test failures are related 😢

**Comment by alesapin** [comment]
I'm 99% that they not. Let's look one more time and investigate this failure tomorrow.

**Comment by alesapin** [comment]
Part of changes https://github.com/ClickHouse/ClickHouse/pull/52498/files

**Comment by alesapin** [comment]
Suspicious integration tests failures:
https://s3.amazonaws.com/clickhouse-test-reports/51851/d1fa105d602f8911263ae4e1f0d4e5f9b0bd15c7/integration_tests__tsan__[5_6].html
https://s3.amazonaws.com/clickhouse-test-reports/51851/d1fa105d602f8911263ae4e1f0d4e5f9b0bd15c7/integration_tests__asan__[2_6].html
https://s3.amazonaws.com/clickhouse-test-reports/51851/6416fb6eed8dc0479735dd63c23379d8206e6edf/integration_tests__release__[2_4].html
Almost not suspicious failure:
https://s3.amazonaws.com/clickhouse-test-reports/51851/6416fb6eed8dc0479735dd63c23379d8206e6edf/integration_tests__release__[1_4].html

100% related stateless tests crash: https://s3.amazonaws.com/clickhouse-test-reports/51851/d1fa105d602f8911263ae4e1f0d4e5f9b0bd15c7/stateless_tests__debug__s3_storage__[3_6].html

**Comment by tavplubix** [comment]
https://s3.amazonaws.com/clickhouse-test-reports/51851/753a8d0a43d94b49b3d310d2406dde247e202c58/integration_tests__tsan__[1_6].html

That was weird:
```
        # node 2 partitioned and don't see any data
>       assert node2.query("SELECT * FROM test_table") == "0\n"
E       AssertionError: assert '0\n33\n2\n' == '0\n'
E           0
E         + 33
E         + 2
```

But the nodes were partitioned at 2023-07-23 22:09:25
```
2023-07-23 22:09:25 [ 632 ] DEBUG : [network] 3b4f1885f5e55540f54f8351bdc194e74ff69af98d886a590c4ef08f5cde418b: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.6.6', '-d', '172.16.6.5', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
2023-07-23 22:09:25 [ 632 ] DEBUG : http://localhost:None "POST /v1.42/containers/3b4f1885f5e55540f54f8351bdc194e74ff69af98d886a590c4ef08f5cde418b/exec HTTP/1.1" 201 74 (connectionpool.py:546, _make_request)
2023-07-23 22:09:25 [ 632 ] DEBUG : http://localhost:None "POST /v1.42/exec/9315d8917e338b44ac42ce311166520dec428715918a4562b2a5762128407de5/start HTTP/1.1" 101 0 (connectionpool.py:546, _make_request)
2023-07-23 22:09:25 [ 632 ] DEBUG : http://localhost:None "GET /v1.42/exec/9315d8917e338b44ac42ce311166520dec428715918a4562b2a5762128407de5/json HTTP/1.1" 200 460 (connectionpool.py:546, _make_request)
2023-07-23 22:09:25 [ 632 ] DEBUG : [network] 3b4f1885f5e55540f54f8351bdc194e74ff69af98d886a590c4ef08f5cde418b: ['iptables', '--wait', '-I', 'DOCKER-USER', '1', '-p', 'tcp', '-s', '172.16.6.5', '-d', '172.16.6.6', '-j', 'DROP'] (0):  (network.py:308, _exec_run)
```
```
3481:time="2023-07-23T22:09:25.259115446Z" level=debug msg="form data: {\"AttachStderr\":true,\"AttachStdin\":false,\"AttachStdout\":true,\"Cmd\":[\"iptables\",\"--wait\",\"-I\",\"DOCKER-USER\",\"1\",\"-p\",\"tcp\",\"-s\",\"172.16.6.6\",\"-d\",\"172.16.6.5\",\"-j\",\"DROP\"],\"Container\":\"3b4f1885f5e55540f54f8351bdc194e74ff69af98d886a590c4ef08f5cde418b\",\"Privileged\":true,\"Tty\":false,\"User\":\"\"}"
3520:time="2023-07-23T22:09:25.292112898Z" level=debug msg="form data: {\"AttachStderr\":true,\"AttachStdin\":false,\"AttachStdout\":true,\"Cmd\":[\"iptables\",\"--wait\",\"-I\",\"DOCKER-USER\",\"1\",\"-p\",\"tcp\",\"-s\",\"172.16.6.5\",\"-d\",\"172.16.6.6\",\"-j\",\"DROP\"],\"Container\":\"3b4f1885f5e55540f54f8351bdc194e74ff69af98d886a590c4ef08f5cde418b\",\"Privileged\":true,\"Tty\":false,\"User\":\"\"}"
```

But node2 managed to download 2 parts somehow:
```
417:2023.07.23 22:09:26.036910 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Fetching part all_10_10_0 from default:/test/table/replicas/r0
418:2023.07.23 22:09:26.037244 [ 226 ] {} <Trace> ReadWriteBufferFromHTTP: Sending request to http://node1:9009/?endpoint=DataPartsExchange%3A%2Ftest%2Ftable%2Freplicas%2Fr0&part=all_10_10_0&client_protocol_version=8&compress=false
424:2023.07.23 22:09:26.039181 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784) (Fetcher): Downloading part all_10_10_0 onto disk default.
425:2023.07.23 22:09:26.040999 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784) (Fetcher): Download of part all_10_10_0 onto disk default finished.
426:2023.07.23 22:09:26.041183 [ 226 ] {} <Trace> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Renaming temporary part tmp-fetch_all_10_10_0 to all_10_10_0 with tid (1, 1, 00000000-0000-0000-0000-000000000000).
427:2023.07.23 22:09:26.041278 [ 226 ] {} <Test> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): preparePartForCommit: inserting all_10_10_0 (state PreActive) into data_parts_indexes
428:2023.07.23 22:09:26.041302 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Committing part all_10_10_0 to zookeeper
429:2023.07.23 22:09:26.044409 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Part all_10_10_0 committed to zookeeper
430:2023.07.23 22:09:26.044537 [ 226 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Fetched part all_10_10_0 from default:/test/table/replicas/r0
```

The iptables rules actually took effect later:
```
2023.07.23 22:09:26.127135 [ 208 ] {} <Debug> default.test_table (2afb5b3d-cdb6-4559-b672-b77186177784): Fetching part all_14_49_1 from default:/test/table/replicas/r0
2023.07.23 22:09:26.127538 [ 208 ] {} <Trace> ReadWriteBufferFromHTTP: Sending request to http://node1:9009/?endpoint=DataPartsExchange%3A%2Ftest%2Ftable%2Freplicas%2Fr0&part=all_14_49_1&client_protocol_version=8&compress=false
2023.07.23 22:09:26.623215 [ 223 ] {} <Trace> HTTPSessionAdapter: Failed to create connection with 172.16.6.6:9009, Will try another resolved address. Poco::Exception. Code: 1000, e.code() = 0, Timeout: connect timed out: 172.16.6.6:9009 (version 23.7.1.1)
2023.07.23 22:09:26.679753 [ 225 ] {} <Trace> HTTPSessionAdapter: Failed to create connection with 172.16.6.6:9009, Will try another resolved address. Poco::Exception. Code: 1000, e.code() = 0, Timeout: connect timed out: 172.16.6.6:9009 (version 23.7.1.1)
```

**Comment by tavplubix** [comment]
Stateless tests flaky check (asan) - `Test runs too long` - okay
Upgrade check (tsan) - https://pastila.nl/?024c2ef6/2eb275edb9d8569304819d60c94d2397, will fix it

**Pull request merged**